pkg/controllers/user/alert/configsyncer/configsyncer.go

```diff
@@ -0,0 +1,252 @@
+package configsyner
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/rancher/norman/controller"
+	alertconfig "github.com/rancher/rancher/pkg/controllers/user/alert/config"
+	"github.com/rancher/rancher/pkg/controllers/user/alert/manager"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func NewConfigSyncer(cluster *config.UserContext, manager *manager.Manager) *ConfigSyncer {
+	return &ConfigSyncer{
+		secretClient:       cluster.Core.Secrets("cattle-alerting"),
+		clusterAlertLister: cluster.Management.Management.ClusterAlerts(cluster.ClusterName).Controller().Lister(),
+		projectAlertLister: cluster.Management.Management.ProjectAlerts("").Controller().Lister(),
+		notifierLister:     cluster.Management.Management.Notifiers(cluster.ClusterName).Controller().Lister(),
+		clusterName:        cluster.ClusterName,
+		alertManager:       manager,
+	}
+
+}
+
+type ConfigSyncer struct {
+	secretClient       v1.SecretInterface
+	projectAlertLister v3.ProjectAlertLister
+	clusterAlertLister v3.ClusterAlertLister
+	notifierLister     v3.NotifierLister
+	clusterName        string
+	alertManager       *manager.Manager
+}
+
+func (d *ConfigSyncer) ProjectSync(key string, alert *v3.ProjectAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) ClusterSync(key string, alert *v3.ClusterAlert) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) NotifierSync(key string, alert *v3.Notifier) error {
+	return d.sync()
+}
+
+func (d *ConfigSyncer) sync() error {
+	//logrus.Info("start sync config")
+
+	clusterAlerts, err := d.clusterAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating cluster alerts")
+	}
+
+	projectAlerts, err := d.projectAlertLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating project alerts")
+	}
+
+	pAlerts := []*v3.ProjectAlert{}
+	for _, alert := range projectAlerts {
+		if controller.ObjectInCluster(d.clusterName, alert) {
+			pAlerts = append(pAlerts, alert)
+		}
+	}
+
+	notifiers, err := d.notifierLister.List("", labels.NewSelector())
+	if err != nil {
+		return errors.Wrapf(err, "Creating project alerts")
+	}
+
+	config := d.alertManager.GetDefaultConfig()
+	config.Global.PagerdutyURL = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+
+	d.addClusterAlert2Config(config, clusterAlerts, notifiers)
+	d.addProjectAlert2Config(config, pAlerts, notifiers)
+
+	data, err := yaml.Marshal(config)
+	//logrus.Infof("after updating notifier: %s", string(data))
+	if err != nil {
+		return errors.Wrapf(err, "Marshal secrets")
+	}
+
+	configSecret, err := d.secretClient.Get("alertmanager", metav1.GetOptions{})
```

Comment:
any reason why cache Get is not used here?

Reply:
okay, I will use Lister.Get here instead

Reply:
Alena, I changed to use Lister.Get here, but it returns an error saying no resource can be found. Michelia also hit this issue. I need to look into the generated code to triage this issue.